src/backend/distributed/master/master_repair_shards.c

```diff
@@ -238,45 +313,282 @@ RecreateTableDDLCommandList(Oid relationId)
+
+	workerConnection = ConnectToNode(nodeName, nodePort, nodeUser);
+	if (workerConnection == NULL)
 	{
-		copySuccessful = true;
+		ereport(ERROR, (errmsg("could not open connection to %s:%d as %s",
+							   nodeName, nodePort, nodeUser)));
 	}
 
-	return copySuccessful;
+	/* start the transaction on the worker node */
+	queryResult = PQexec(workerConnection, "BEGIN");
+	if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+	{
+		ReraiseRemoteError(workerConnection, queryResult);
+	}
+
+	PQclear(queryResult);
+
+	/* iterate over the commands and execute them in the same connection */
+	foreach(commandCell, commandList)
+	{
+		char *commandString = lfirst(commandCell);
+		ExecStatusType resultStatus = PGRES_EMPTY_QUERY;
+
+		queryResult = PQexec(workerConnection, commandString);
+		resultStatus = PQresultStatus(queryResult);
+		if (!(resultStatus == PGRES_SINGLE_TUPLE || resultStatus == PGRES_TUPLES_OK ||
+			  resultStatus == PGRES_COMMAND_OK))
+		{
+			ReraiseRemoteError(workerConnection, queryResult);
+		}
+
+		PQclear(queryResult);
+	}
+
+	/* commit the transaction on the worker node */
+	queryResult = PQexec(workerConnection, "COMMIT");
+	if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+	{
+		ReraiseRemoteError(workerConnection, queryResult);
+	}
+
+	PQclear(queryResult);
+
+	/* clear NULL result */
+	PQgetResult(workerConnection);
+
+	/* we no longer need this connection */
+	PQfinish(workerConnection);
+}
+
+
+/*
+ * ConstuctQualifiedShardName creates the fully qualified name string of the
+ * given shard in <schema>.<table_name>_<shard_id> format.
+ *
+ * FIXME: Copied from Citus-MX, should be removed once those changes checked-in to Citus.
+ */
+static char *
+ConstructQualifiedShardName(ShardInterval *shardInterval)
+{
+	Oid schemaId = get_rel_namespace(shardInterval->relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *tableName = get_rel_name(shardInterval->relationId);
+	char *shardName = NULL;
+
+	shardName = pstrdup(tableName);
+	AppendShardIdToName(&shardName, shardInterval->shardId);
+	shardName = quote_qualified_identifier(schemaName, shardName);
+
+	return shardName;
+}
+
+
+/*
+ * CitusExtensionOwnerName returns the name of the owner of the extension.
+ *
+ * FIXME: Copied from Citus-MX, should be removed once those changes checked-in to Citus.
+ */
+static char *
+CitusExtensionOwnerName(void)
+{
+	Oid superUserId = CitusExtensionOwner();
+
+#if (PG_VERSION_NUM < 90500)
+	return GetUserNameFromId(superUserId);
+#else
+	return GetUserNameFromId(superUserId, false);
+#endif
+}
+
+
+/*
+ * DropShardPlacement deletes the shard placement metadata for the given shard
+ * placement from the pg_dist_shard_placement, and then it drops the shard table
+ * from the given node.
+ *
+ * FIXME: Copied from Citus-MX, should be removed once those changes checked-in to Citus.
+ */
+static void
+DropShardPlacement(ShardInterval *shardInterval, char *nodeName, int32 nodePort)
+{
+	StringInfo dropQuery = NULL;
```

Comment:
Well, I'm thinking that we should probably do not allow to drop a single shard. Instead, as in copy, we should find the co-located ones here, and drop all of them. Otherwise, we might end-up with broken co-location and lead to failing queries / wrong results etc.

what do you think?